src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -181,12 +187,225 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+
+  // Use projection to calculate query's distance from hash limits
+  arma::vec limLow = queryCodeNotFloored - projection;
+  arma::vec limHigh = hashWidth - limLow;
+
+  // calculate scores = distances^2
+  arma::vec scores(2 * numProj);
+  scores.rows(0, numProj - 1) = arma::pow(limLow, 2);
+  scores.rows(numProj, 2 * numProj - 1) = arma::pow(limHigh, 2);
+
+  // actions vector shows what transformation to apply to a coordinate
+  arma::Col<short int> actions(2 * numProj); //will be [-1 ... 1 ...]
+
+  actions.rows(0, numProj - 1) = //first numProj rows
+    -1 * arma::ones< arma::Col<short int> > (numProj); //-1s
+
+  actions.rows(numProj, 2 * numProj - 1) = //last numProj rows
+    arma::ones< arma::Col<short int> > (numProj); //1s
+
+
+  // acting dimension vector shows which coordinate to transform according to
+  // actions described by actions vector
+  arma::Col<size_t> positions(2 * numProj); //will be [0 1 2 ... 0 1 2 ...]
+  positions.rows(0, numProj - 1) =
+    arma::linspace< arma::Col<size_t> >(0, numProj - 1, numProj);
+  positions.rows(numProj, 2 * numProj - 1) =
+    arma::linspace< arma::Col<size_t> >(0, numProj - 1, numProj);
+
+  // sort everything in increasing order
+  arma::Col<long long unsigned int> sortidx = arma::sort_index(scores);
+  scores = scores(sortidx);
+  actions = actions(sortidx);
+  positions = positions(sortidx);
+
+
+  // Theory:
+  // From the paper: This is the part that creates the probing sequence
+  // A probing sequence is a sequence of T probing bins where query's
+  // neighbors are most likely to be. Likelihood is dependent only on a bin's
+  // score, which is the sum of scores of all dimension-action pairs, so we
+  // need to calculate the T smallest sums of scores that are not conflicting.
+  //
+  // Method:
+  // Store each perturbation set (pair of (dimension, action)) in a
+  // std::vector. Create a minheap of scores, with each node pointing to its
+  // relevant perturbation set. Each perturbation set popped from the minheap
+  // is the next most likely perturbation set.
+  // Transform perturbation set to perturbation vector by setting the
+  // dimensions specified by the set to queryCode+action (action is {-1, 1}).
+
+  std::vector<size_t> Ao;
+  Ao.push_back(0); // initial perturbation holds smallest score (0 if sorted)
+
+  std::vector< std::vector<size_t> > perturbationSets;
+  perturbationSets.push_back(Ao); //storage of perturbation sets
+
+  //define a priority queue with CompareGreater as a minheap
+  std::priority_queue<
+    std::pair<double, size_t>,        //contents: pairs of (score, index)
+    std::vector<                      //container: vector of pairs
+      std::pair<double, size_t>
+      >,
+    mlpack::neighbor::CompareGreater  //comparator of pairs (compares scores)
+  > minHeap; //our minheap
+
+  //Start by adding the lowest scoring set to the minheap
+  std::pair<double, size_t> pair0( perturbationScore(Ao, scores), 0 );
```

Comment:
priority_queue is a min-heap and the default comparator is defined for pair, so if you just invert your scores here so you want the min instead of the max I think it will allow you to remove a lot of code, including the custom comparator.